src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -334,6 +360,62 @@ case class FilterCols(
   }
 }
 
+// FilterColsIR is used when the predicate passes toIR()
+
+case class FilterColsIR(
+  child: MatrixIR,
+  pred: IR) extends MatrixIR {
+
+  def children: IndexedSeq[BaseIR] = Array(child, pred)
+
+  def copy(newChildren: IndexedSeq[BaseIR]): FilterColsIR = {
+    assert(newChildren.length == 2)
+    FilterColsIR(newChildren(0).asInstanceOf[MatrixIR], newChildren(1).asInstanceOf[IR])
+  }
+
+  def typ: MatrixType = child.typ
+
+  def execute(hc: HailContext): MatrixValue = {
+    val prev = child.execute(hc)
+
+    val localGlobals = prev.globals.broadcast
+    val localColType = typ.colType
+    val ec = typ.colEC
+	  //
+	  // Initialize a region containing the globals
+	  //
+	  val colRegion = Region()
+	  val rvb = new RegionValueBuilder(colRegion)
+	  rvb.start(typ.globalType)
+	  rvb.addAnnotation(typ.globalType, localGlobals.value)
+    val globalRVend = rvb.currentOffset()
+	  val globalRVoffset = rvb.end()
+
+    val (rTyp, predCompiledFunc) = ir.Compile[Long, Long, Boolean](
+      "global", typ.globalType,
+      "sa",     typ.colType,
+      pred
+    )
+	  //
+	  // Hmm ... is this going to do a separate reduction for each column/sample ?
+	  // Won't that be slow in the distributed case ?
+	  //
+    val colAggregationOption = Aggregators.buildColAggregations(hc, prev, ec)
+
+    val p = (sa: Annotation, i: Int) => {
+      colAggregationOption.foreach(f => f.apply(i))
+      ec.setAll(localGlobals, sa)
+      colRegion.clear(globalRVend)
+      val colRVb = new RegionValueBuilder(colRegion)
+      colRVb.start(localColType)
+      colRVb.addAnnotation(localColType, sa)
+      val colRVoffset = colRVb.end()
+      predCompiledFunc()(colRegion, globalRVoffset, false, colRVoffset, false) == true
```

Comment:
The `== true` isn't necessary anymore because the function returns Boolean, not java.lang.Boolean (that is, it can't be null)

Reply:
done